odl/discr/discr_mappings.py

```diff
@@ -687,42 +790,185 @@ def _evaluate(self, indices, norm_distances, out=None):
+        low_weights.append(w_lo)
+        high_weights.append(w_hi)
+        edge_indices.append(edge)
 
+    return low_weights, high_weights, edge_indices
 
-class _LinearPointwiseInterpolator(_PointwiseInterpolator):
 
-    """Linear interpolator for point arrays.
+class _PerAxisInterpolator(_Interpolator):
 
-    The code is adapted from SciPy's `RegularGridInterpolator
-    <http://docs.scipy.org/doc/scipy/reference/generated/\
-scipy.interpolate.RegularGridInterpolator.html>`_ class.
+    """Interpolator where the scheme is set per axis.
+
+    This allows to use e.g. nearest neighbor interpolation in the
+    first dimension and linear in dimensions 2 and 3.
     """
 
+    def __init__(self, coord_vecs, values, typ, schemes, nn_variants):
+        """Initialize a new instance.
+
+        coord_vecs : `tuple` of `numpy.ndarray`
+            Coordinate vectors defining the interpolation grid
+        values : array-like
+            Grid values to use for interpolation
+        typ : {'array', 'meshgrid'}
+            Type of expected input values in ``__call__``
+        schemes : `list` of strings
+            Indicates which interpolation scheme to use for which axis
+        nn_variants : `list` of strings
+            Which variant ('left' or 'right') to use in nearest neighbor
+            interpolation for which axis.
+            This option has no effect for schemes other than nearest
+            neighbor.
+        """
+        super().__init__(coord_vecs, values, typ)
+        self.schemes = schemes
+        self.nn_variants = nn_variants
+
     def _evaluate(self, indices, norm_distances, out=None):
-        """Evaluate nearest interpolation. Modified for in-place."""
+        """Evaluate linear interpolation.
+
+        Modified for in-place evaluation and treatment of out-of-bounds
+        points by implicitly assuming 0 at the next node."""
         # slice for broadcasting over trailing dimensions in self.values
         vslice = (slice(None),) + (None,) * (self.values.ndim - len(indices))
 
-        # find relevant values
-        # each i and i+1 represents a edge
-        edges = product(*[[i, i + 1] for i in indices])
-        values = 0.
-        for edge_indices in edges:
-            weight = 1.
-            for ei, i, yi in zip(edge_indices, indices, norm_distances):
-                weight *= np.where(ei == i, 1 - yi, yi)
-            values += np.asarray(self.values[edge_indices]) * weight[vslice]
-        return values
+        out_shape = out_shape_from_meshgrid(norm_distances)
+        out_dtype = norm_distances[0].dtype
+
+        if out is None:
+            out = np.zeros(out_shape, dtype=out_dtype)
+        else:
+            out[:] = 0.0
+
+        # Weights and indices (per axis)
+        low_weights, high_weights, edge_indices = _create_lists(
+            indices, norm_distances, self.schemes, self.nn_variants)
+
+        # Iterate over all possible combinations of [i, i+1] for each
```

Comment:
Sure. This should be robust, however, since it's equivalent to [the original scipy inmplementation](https://github.com/scipy/scipy/blob/master/scipy/interpolate/interpolate.py#L1773-L1786).